docs: adds oauth plain authentication flows (#6262)
strimzi/strimzi-kafka-operator: documentation/modules/oauth/con-oauth-authentication-client-options.adoc

// assembly-oauth-authentication.adoc

[id='con-oauth-authentication-client-options-{context}']
= OAuth 2.0 client authentication flows

[role="_abstract"]
OAuth 2.0 authentication flows depend on the underlying Kafka client and Kafka broker configuration.
The flows must also be supported by the authorization server used.

Kafka broker listener configuration determines how clients authenticate using an access token.
The client can pass a client ID and secret to request an access token.

> If a listener is configured to use PLAIN authentication, you can use a client username and password to request or pass access tokens.

Reviewer: This doesn't sound right. there is no such thing as 'client username'. There is either `clientId` or `username`. The way it sounds here is as if we support password authentication, where a user can configure the client with some user account's username and password. That is not supported even when PLAIN authentication is used. What is supported is to authenticate with clientId and secret or with username and access token and passing these as `username` and `password` parameters of the SASL/PLAIN authentication mechanism.  
```suggestion
If a listener is configured to use PLAIN authentication, PLAIN mechanism can be used on the client to authenticate with `clientId` and `secret`, or `username` and `access token` by passing these as `username` and `password` parameters of the PLAIN mechanism.
```

Author: I've updated as follows:

> If a listener is configured to use PLAIN authentication, the client can authenticate with a client ID and secret or username and access token. These values are passed as the `username` and `password` properties of the PLAIN mechanism.

Reviewer: Sounds better, yes.

Listener configuration supports the following token validation options:

* Fast local token validation based on JWT signature checking and local token introspection, without contacting an authorization server.
The authorization server provides a JWKS endpoint with public certificates that are used to validate signatures on the tokens.
* A call to a token introspection endpoint provided by an authorization server.
Each time a new Kafka broker connection is established, the broker passes the access token received from the client to the authorization server.
The Kafka broker checks the response to confirm whether or not the token is valid.

Kafka client credentials can also be configured for the following types of authentication:

* Direct local access using a previously generated long-lived access token

> * Contact with the authorization server for a new access token to be issued

Reviewer: Maybe add here a bit more detail?
```suggestion
* Contact with the authorization server for a new access token to be issued (by using `clientId` and a `secret` or a `refresh token`)
```

> NOTE: An authorization server might only allow the use of opaque access tokens, which means that local token validation is not possible.

Reviewer: I would move this after line 23.

== Example client authentication flows using the SASL OAUTHBEARER mechanism

> You can use the following communication flows for Kafka authentication using the OAuth SASL OAUTHBEARER mechanism.

Reviewer: ```suggestion
You can use the following communication flows for Kafka authentication using the SASL OAUTHBEARER mechanism.
```

// assembly-oauth-authentication.adoc

[id='con-oauth-authentication-client-options-{context}']
= OAuth 2.0 client authentication flows

[role="_abstract"]
OAuth 2.0 authentication flows depend on the underlying Kafka client and Kafka broker configuration.
The flows must also be supported by the authorization server used.

> Kafka broker listener configuration determines how clients authenticate using an access token.

Reviewer 2: The Kafka broker listener ...

The client can pass a client ID and secret to request an access token.

If a listener is configured to use PLAIN authentication, the client can authenticate with a client ID and secret or username and access token.
These values are passed as the `username` and `password` properties of the PLAIN mechanism.

Listener configuration supports the following token validation options:

> * Fast local token validation based on JWT signature checking and local token introspection, without contacting an authorization server.

Reviewer 2: Refrain from using fragments.

Maybe try...

You can use fast local token validation based on JWT signature checking and local token introspection, without contacting an authorization server.

The authorization server provides a JWKS endpoint with public certificates that are used to validate signatures on the tokens.

> * A call to a token introspection endpoint provided by an authorization server.

Reviewer 2: Please rewrite so it is not an incomplete sentence.

Token expiration is written into the token, but revocation can happen at any time, so cannot be accounted for without contacting the authorization server.
Any issued token would be considered valid until it expires.

== Example client authentication flows using the SASL PLAIN mechanism

You can use the following communication flows for Kafka authentication using the OAuth PLAIN mechanism.

* xref:oauth-plain-client-id-{context}[Client using a client ID and secret, with the broker obtaining the access token for the client]
* xref:oauth-plain-access-token-{context}[Client using a long-lived access token without a client ID and secret]

[id='oauth-plain-client-id-{context}']
.Client using a client ID and secret, with the broker obtaining the access token for the client

image:oauth-plain-client-id.png[Client using a client ID and secret, with the broker obtaining the access token for the client]

. The Kafka client passes a `clientId` as a username and a `secret` as a password.
. The Kafka broker uses a token endpoint to pass the `clientId` and `secret` to the authorization server.

> . The authorization server returns a fresh access token in exchange or an error if the client credentials aren't valid.

Reviewer 2: please remove 'in exchange'

Also, I think it is safer to not use the contraction aren't here. I know the style guide has updated rules on this, but it still poses some issues with translation and localization.

. The Kafka broker validates the token in one of the following ways:
.. If a token introspection endpoint is specified, the Kafka broker validates the access token by calling the endpoint on the authorization server.

> A session is established if the token validation was successful.

Reviewer 2: I think 

A session is established if the token validation is successful.

> .. If local token introspection is used, there is no request made to the authorization server.

Reviewer 2: I think a possible rewrite. 

If local token introspection is used, a request is not made to the authorization server.

image:oauth-plain-access-token.png[Client using a long-lived access token without a client ID and secret]

. The Kafka client passes a username and password. The password provides the value of an access token that was obtained manually and configured before running the client.
. The password is passed with or without an `$accessToken:` string prefix depending on whether or not the Kafka broker listener is configured with a token endpoint for authentication.
.. If the token endpoint is configured, the password should be prefixed by `$accessToken:` to let the broker know that the password parameter contains an access token rather than a client secret. The Kafka broker interprets the username as the account username.
.. If the token endpoint is not configured on the Kafka broker listener (enforcing a `no-client-credentials mode`), the password should provide the access token without the prefix. The Kafka broker interprets the username as the account username.
In this mode, the client doesn't use a client ID and secret, and the `password` parameter is always interpreted as a raw access token.
. The Kafka broker validates the token in one of the following ways:

> .. If a token introspection endpoint is specified, the Kafka broker validates the access token by calling the endpoint on the authorization server. Session is established if token validation was successful.

Reviewer 2: ```suggestion
If a token introspection endpoint is specified, the Kafka broker validates the access token by calling the endpoint on the authorization server. A session is established if token validation is successful.
```
